README.md
Rehost -> DTFE 1.2B
==================

This is a more stable but unofficial version of DTFE 1.2 (see the change log). Most notably there have been a few bug fixes in the file IO. The owner of this repository is not affiliated with the original authors of DTFE. The following text was taken directly from the [DTFE](http://www.astro.rug.nl/~voronoi/DTFE/dtfe.html) website.

The DTFE method
==================

![](./docs/steps.png)

The Delaunay Tessellation Field Interpolation (DTFE) method represents the natural way of going from discrete samples/measurements to values on a periodic grid using the maximum of information contained in the points set (that is the measurements points or particle distribution).

The DTFE method is especially suitable for astronomical data due to the following reasons:

 - Preserves the multi-scale character of the point distribution. This is the case in numerical simulations of large scale structure (LSS) where the density varies over more than 6 orders of magnitude.
 - Preserves the local geometry of the point distribution. This is important in recovering sharp features like the different components of the LSS (i.e. clusters, filaments, walls and voids).
 - DTFE does not depend on user defined parameters or choices.
 - The interpolated fields are volume weighted (versus mass weighted quantities in most other interpolation schemes). This can have a significant effect especially when comparing with analytical predictions which are volume weighted.

For detailed information about the DTFE method see the [Papers](http://www.astro.rug.nl/~voronoi/DTFE/papers.html) section.

The DTFE public software
==================

DTFE is free software, distributed under the [GNU General Public License](http://www.gnu.org/copyleft/gpl.html). This implies that you may freely distribute and copy the software. You may also modify it as you wish, and distribute these modified versions as long as you indicate prominently any changes you made in the original code, and as long as you leave the copyright notices, and the no-warranty notice intact. Please read the General Public License for more details. Note that the authors retain their copyright on the code.

The DTFE public code is a C++ implementation of the DTFE grid interpolation method. The code was written by Marius Cautun at the Kapteyn Astronomical Institute, Netherlands, with the purpose of analyzing cosmological numerical simulations and galaxy redshift survey. Even though the code was designed with astrophysics in mind, it can be used for problems in a wide range of fields where one needs to interpolate from a discrete set of points to a grid.

The code was designed using a modular philosophy and with a wide set of features that can easily be selected using the different program options. The DTFE code is also written using OpenMP directives which allow it to run in parallel on shared-memory architectures.

The code comes with a complete documentation and with a multitude of examples that detail the program features. Moreover, a help desk is available for information and assistance for troubleshooting problems.


Current features in the DTFE code
==================

 - Works in both 2 and 3 spatial dimensions.
 - Interpolates the fields to three different grids:
    - Regular rectangular grid.
    - Redshift cone grid.
    - User given sampling points - these can describe any complex geometry.
 - Uses the point distribution itself to compute the density and interpolates the result to grid.
 - Each sample point has a weight associated to it to represent multiple resolution N-body simulations and observational biases for galaxy redshift surveys.
 - Interpolates the velocity, velocity gradient, velocity divergence, velocity shear and velocity vorticity.
 - Interpolates any additional number of fields and their gradients to grid.
 - Periodic boundary conditions.
 - Zoom in option for regions of interest.
 - Splitting the full data in smaller computational chunks when dealing with limited CPU resources.
 - The computation can be distributed in parallel on shared-memory architectures.
 - Also comes with the TSC and the SPH interpolation methods.
 - Can return the Delaunay tessellation of the given point set.
 - Easy change of input/output data format.
 - Easy to use as an external library.
 - Extensive documentation of each feature.

At the moment the code comes with a Gadget snapshot reader for the input data, as well as with examples on how to load data from binary and text files. There are plans to offer input data readers for other types of input files (especially those specific to N-body codes), so if you are using the DTFE program with another reader, you can contact the help desk if you would like to include that reader in the official distribution. At the same time we are also looking for output data writing functions (especially to formats used in analysis and visualization programs) - at the moment the programs writes the output data to a binary or text file.


change_log.txt
VERSION 1.2B
(version modified by con-f-use. Known issues: 3D-scalars, e.g. acceleration is not read out yet)
- fixed minor bug causing segmentation faults when reading gadget snapshots
- fixed an issue with skipping blocks of data in gadget snapshots
- the gadget reader now checks the block delimeter (size of block) for every block as opposed to just the header
- implemented better handling of scalar fields for gadget snapshots: the `--input`-option now knows one additional data type to read:
      8 - internal energy
     16 - unknown other block (if internal energy is not present - else this will also be dedicated to internal energy and the other block will be 32)
     32 - unknown other block (if internal energy is present and not read read by user option)
  so `--input 101 8 1` would now map the internal energy to an output file. Additional blocks after that can be read. Internal energy is a special block as its value must be multiplied by the mass. If you don't read internal energy but a later block the later block will usually correspond to 16. That is if no internal energy block is present. If an internal energy block is present and not read by user option, a later block will correspond to 32 (not 16) since internal energy will be read (skipped) by the program as 16 "unknown gas block".
- edited the `--full_help`-option to reflect the changes
- fixed error in the binary io, all data was read into the positions array, also velocities are not mandatory anymore


VERSION 1.2
Added some new features:
- added code for reading data from an HDF5 Gadget file

- added a new option called '--input' or '-i' that can be used to specify the input file type, the data to read from the input file and for what species. The '--input' option has the form: '--input  input_file_type   data_to_read    particle_species_to_read'   where the last two make sense only for Gadget files. The DEFAULT values are: '--input 101 7 2'
- the new option '--input' or '-i' needs to be followed by an integer specifying the input file type. By default it understands the following options:
      101 - gadget binary snapshot saved as single or multiple files type I or II [ DEFAULT choice unsless specified otherwise in the Makefile ]
      105 - gadget HDF5 file/files
      111 - reads particles positions and masses (weights) from a text file. First line gives particle number followed by 6 coordinates giving the box dimensions: "xMin xMax yMin yMax zMin zMax". After this each row 'i' in the file gives positions X, Y, Z and weight for particle 'i'.
      112 - reads particles positions from a text file. First line gives particle number followed by 6 coordinates giving the box dimensions: "xMin xMax yMin yMax zMin zMax". After this each row 'i' in the file gives positions X, Y and Z for particle 'i'. (note that each particle is assumed to have the same mass/weight)
      121 - read particle positions, mass and velocities from a binary file. Modify this function in file 'src/io/binary_io.cc'
      131 - read particle positions, mass and velocities from your custom file. Modify this function in file 'src/io/my_io.cc'
you can set the default output file in the Makefile by changing the values assigned to the '-DINPUT_FILE_DEFAULT' option.
- added a second integer value to the '--input' option to specify what which data to read. At the moment works only for gadget files and you can select to read the following:
      1 - read particle positions
      2 - read particle weights
      4 - read particle velocities
to read multiple data add the number corresponding to each data set, for example 3=1+2 - specifies to read only particle positions and weights/masses, while a value of 7=1+2+4 specifies to read positions, weights and velocities [ DEFAULT = 7 -read positions, weights and velocities ]
- added an additional third option after for the '--input' parameter which specifies for which gadget particle species to read the data. You can select the following:
      1 - read the data for gadget species 1 (normally gas particles)
      2 - read the data for gadget species 2 (normally dark matter particles) [DEFAULT]
      4 - read the data for gadget species 3 (usually stars or different mass DM particles in multiple resolution simulations)
      8 - read the data for gadget species 4
      16 - read the data for gadget species 5
      32 - read the data for gadget species 6
to read data for multiple species just add the number corresponding to each particle type, for example 7=1+2+4 - specifies to read the data for both particle types 1, 2 and 3.

- added the '--output' or '-o' option to specify the output file type. Available options:
      101 - the default binary output
      100 - binary file with my own custom header and file structure
you can set the default output file in the Makefile by changing the values assigned to the '-DOUTPUT_FILE_DEFAULT' option.

- added option '--resdhiftSpace' that can be used to tranform the particle positions from position-space to redshift-space. For the option to work you need to enable the '-DREDSHIFT_SPACE' option in the Makefile. Than you need to call the '--resdhiftSpace d1 d2 d3' during runtime with (d1,d2,d3) the direction along which to use the velocity for the transformation to redshift space. The (d1,d2,d3) direction will be normalized to a unit vector by the program!




VERSION 1.1.2
- fixed bug when using the '--partition nx ny nz' with nx*ny*nz>=27


src/io/binary_io.cc
/*
 *  Copyright (c) 2011       Marius Cautun
 *
 *                           Kapteyn Astronomical Institute
 *                           University of Groningen, the Netherlands
 *
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.al;kjdsl;kadj
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */



/* This file contains functions for reading and writing the data to a binary file. */


/* This function reds the input data from a binary file.
This function works when the binary file has the following format:
    1) file begins with an int value that gives the number of particles
    2) continues with 6 float values that give the coordinates of the box encompassing the data (i.e. xMin, xMax, yMin, yMax, zMin, zMax)
    3) the next part of the file gives the positions of the particles which are saved as x1, y1, z1, x2, y2, z2, x3, ... (with x1 = x coordinate for particle 1, y1 = y coordinate for particle 1, etc ...)
    4) after the particle positions, come the particle weights array saved as w1, w2, w3, ... (with w1 = weight of particle 1, etc ...)
    5) after the weights array, comes the optional velocity array which are is saved as vx1, vy1, vz1, vx2, vy2, vz2, ... (with vx1 = the velocity along the x direction of particle 1, ...) and assumed zero, if not in file

NOTE: In this example the data is saved in the binary file in single precision format (float 32 bytes).
*/
void readBinaryFile(std::string filename,
                    Read_data<float> *readData,
                    User_options *userOptions)
{
    MESSAGE::Message message( userOptions->verboseLevel );
    message << "Reading the input data from the binary file '" << filename << "' ... " << MESSAGE::Flush;

    // open the file
    std::fstream inputFile;
    openInputBinaryFile( inputFile, filename );


    // read the number of particles and the box coordinates from the binary file
    int noParticles;
    float boxCoordinates[2*NO_DIM];
    inputFile.read( reinterpret_cast<char *>(&noParticles), sizeof(noParticles) );
    inputFile.read( reinterpret_cast<char *>(boxCoordinates), 6*sizeof(float) );
    checkFileOperations( inputFile, "read from" );

    for (size_t i=0; i<2*NO_DIM; ++i) {
        userOptions->boxCoordinates[i] = boxCoordinates[i];
    }


    // reserve memory for the input data
    Real *positions = readData->position(noParticles);  //particle positions
    Real *weights = readData->weight(noParticles);      //particle weights (e.g. weights = particle masses)
    Real *velocities = readData->velocity(noParticles); //particle velocities


    // read the rest of the input data: positions, weights and velocities
    size_t dataSize = noParticles * sizeof(float) * NO_DIM;    // number of data bytes that store the particle positions (3*4 bytes per particle)
    inputFile.read( reinterpret_cast<char *>(positions), dataSize );
    checkFileOperations( inputFile, "read from" );   // check that the data reading was successful

    dataSize = noParticles * sizeof(float);    // number of data bytes that store the particle weights (1*4 bytes per particle)
    inputFile.read( reinterpret_cast<char *>(weights), dataSize );
    checkFileOperations( inputFile, "read from" );   // check that the data reading was successful

    dataSize = noParticles * sizeof(float) * NO_DIM;    // number of data bytes that store the particle velocities (3*4 bytes per particle)
    inputFile.read( reinterpret_cast<char *>(velocities), dataSize );
    if( inputFile.eof() ) {
        message << "Done.\n";
        message << "Binary input file '" << filename << "' does not contain velocity data!\n" << MESSAGE::Flush;
        for(int it=0; it<noParticles; ++it) velocities[it] = 0.0; // Set velocities zero for good measure
    }
    else
        checkFileOperations( inputFile, "read from" );   // check that the data reading was successful

    inputFile.close();
}



/* This function writes the data to a binary file.*/
template <typename T>
void writeBinaryFile(T const &dataToWrite,
                     std::string filename,
                     std::string variableName,
                     User_options const &userOptions)
{
    MESSAGE::Message message( userOptions.verboseLevel );
    message << "Writing the " << variableName << " to the binary file '" << filename << "' ...  " << MESSAGE::Flush;

    // open the file
    std::fstream outputFile;
    openOutputBinaryFile( outputFile, filename );


    // write the data to file
    // it is very simple to write data to a binary file, just do: outputFile.write( reinterpret_cast<char const *>(&(dataToWrite[0])), dataSize ). The following implements a recursive method that deals with possible issuess in writing very large data sets.
    size_t dataSize = dataToWrite.size()*sizeof(dataToWrite[0]);    //total number of data bytes to be written
    size_t maxSize = 256*256*256;   // write at most 256^3 elements at once, otherwise the write function might fail
    size_t noRepeats = size_t( dataToWrite.size() / maxSize ), currentPosition = 0;
    size_t tempBuffer = maxSize * sizeof(dataToWrite[0]);
    for (size_t i=0; i<noRepeats; ++i)  //write in blocks of 256^3 elements
    {
        outputFile.write( reinterpret_cast<char const *>(&(dataToWrite[currentPosition])), tempBuffer );
        currentPosition += maxSize;
    }
    tempBuffer = (dataToWrite.size() - currentPosition) * sizeof(dataToWrite[0]);    // write everything else that is left
    outputFile.write( reinterpret_cast<char const *>(&(dataToWrite[currentPosition])), tempBuffer );

    checkFileOperations( outputFile, "write to" );   // check that the data writing was succesful
    outputFile.close();

    message << "Done.\n";
}




src/input_output.cc
/*
 *  Copyright (c) 2011       Marius Cautun
 *
 *                           Kapteyn Astronomical Institute
 *                           University of Groningen, the Netherlands
 *
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/*!
Here are the functions used for data input and output.
To easily find how to read the input data go to the function:



*/

#ifndef INPUT_OUTPUT_HEADER
#define INPUT_OUTPUT_HEADER

#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <cstdlib>
#include <cmath>
#include <vector>
#include <string>
#include <boost/filesystem.hpp>
namespace bfs=boost::filesystem;

#include "define.h"
#include "particle_data.h"
#include "user_options.h"
#include "message.h"


// contains the definitions of some classes and fucntions used only for input and output purposes
#include "io/input_output.h"

// different data format readers and writters
#include "io/gadget_reader.cc"  // reader for Gadget snapshots
#include "io/binary_io.cc"      // writer for binary file format
#include "io/text_io.cc"        // reader/writer for text files
#include "io/my_io.cc"          // reader/writer for text files
#include "io/density_file_io.cc"    // own output file format



void openInputBinaryFile(std::fstream & inputFile,
                         std::string & fileName);
void openOutputBinaryFile(std::fstream & outputFile,
                          std::string & fileName);
void openInputTextFile(std::fstream & inputFile,
                       std::string & fileName);
void openOutputTextFile(std::fstream & outputFile,
                        std::string & fileName);





//! Functions for reading the input data

/* Read only positions from a text file and on top of that also reads user defined sampling points from an additional file (USE ONLY WHEN YOU NEED CUSTOM SAMPLING POINTS) - see the "src/io/text_io.cc" for the definition of this function */
//#define DATA_INPUT_FUNCTION readTextFile_userDefinedSampling

/* Read the input data from a binary file - see the "src/io/binary_io.cc" for the definition of this function */
//#define DATA_INPUT_FUNCTION readBinaryFile


typedef void (*FunctionReadInputData)(std::string, Read_data<float> *, User_options *);

/*! This function uses the 'inputFileType' entry in the userOptions strcuture to decide on the function that will be used to read the input data. */
FunctionReadInputData chooseInputDataReadFunction(int const inputFileType)
{
    if ( inputFileType==101 )
        return &readGadgetFile;    // Read the input data from a single/multiple Gadget snapshot file (works only for snapshots type 1 or 2). See the "src/io/gadget_reader.cc" for the definition of this function.
#ifdef HDF5
    else if ( inputFileType==105 )
        return &HDF5_readGadgetFile;        // Read the input data from a HDF5 Gadget snapshot file. See the "src/io/gadget_reader.cc" for the definition of this function.
#endif
    else if ( inputFileType==111 )
        return &readTextFile;               // Read the input data from a text file. See the "src/io/text_io.cc" for the definition of this function.
    else if ( inputFileType==112 )
        return &readTextFile_positions;     // Read the input data from a text file. The text file has only particle positions. See the "src/io/text_io.cc" for the definition of this function.
    else if ( inputFileType==121 )
        return &readBinaryFile;             // Define your custom function to read the input data. Do this in the "src/io/binary_io.cc" file.
    else if ( inputFileType==131 )
        return &readMyFile;                 // Define your custom function to read the input data. Do this in the "src/io/my_io.cc" file.
    else
        throwError( "Unknow value for the 'inputFileType' argument in function 'chooseInputDataReadFunction'. The program could not recognize the input data file type." );
}




/*! This function reads the particle data used for the DTFE computation. You can modify this function as you please. */
void readInputData(std::vector<Particle_data> *p,
                   std::vector<Sample_point> *samplingCoordinates,
                   User_options *userOptions)
{
    std::string filename = userOptions->inputFilename;  // the name of the input file
    Read_data<Real> readData; // will store the data read from the input file


    // Read the data from the input file - see the function 'chooseInputDataReadFunction' that selects the input function used to read in the input data file
    FunctionReadInputData functionReadInputData = chooseInputDataReadFunction( userOptions->inputFileType );
    (*functionReadInputData)( filename, &readData, userOptions );


    // 'userOptions->MpcValue' is the conversion factor from the units in the input data file to Mpc units - do the next computation only if userOptions->MpcValue!=1
    if ( userOptions->MpcValue!=Real(1.) )
    {
        for (size_t i=0; i<userOptions->boxCoordinates.size(); ++i)
            userOptions->boxCoordinates[i] /= userOptions->MpcValue;

        size_t noParticles = readData.noParticles();  // returns the number of particles
        float *positions = readData.position();  //returns pointer to array storing particle positions
        for (size_t i=0; i<noParticles*NO_DIM; ++i)
            positions[i] /= userOptions->MpcValue;

        // if the user inserted user defined sampling points, divide the coordinates of those points by the normalization factor
        if ( readData.noSamples()!=size_t(0) )
        {
            float *sampling = readData.sampling();  // returns pointer to user defined sampling coordinates
            float *delta = readData.delta();        // returns pointer to user defined sampling cell sizes
            for (size_t i=0; i<readData.noSamples()*NO_DIM; ++i)
            {
                sampling[i] /= userOptions->MpcValue;
                delta[i] /= userOptions->MpcValue;
            }
        }
    }


    // now store the data in the 'Particle_data list'. It also copies the user given sampling coordinates, if any - none in this case.
    redshiftSpaceOn = userOptions->transformToRedshiftSpaceOn;
    redshiftSpaceVector = Pvector<Real,NO_DIM>( &(userOptions->transformToRedshiftSpace[0]) );

    readData.transferData( p, samplingCoordinates );
    //std::cout << "\nFirst 10 scalar values transfered\n";
    //for (int i=0; i<10; ++i) std::cout << "\tscalar: " << (*p).at(i).scalar(0) << "\n";
}










//! Functions for writing the output data


//! short class to easily change between writing the output to different file types while the program is running. The user can choose the output file type using the '--output' option.
class OutputData
{
    public:
    int outputFileType;

    // class constructor - initializes the type of the output file
    OutputData(int fileType)
    {
        this->outputFileType = fileType;
    }

    // the function that calls the function doing the actual writing
    template <typename T>
    void write(T &dataToWrite,
                std::string filename,
                std::string variableName,
                User_options const &userOptions)
    {
        if ( this->outputFileType==101 )
            writeBinaryFile( dataToWrite, filename, variableName, userOptions );        // writes the data to a binary file (see "binary_io.cc" for function definition)
        else if ( this->outputFileType==111 )
            writeTextFile( dataToWrite, filename, variableName, userOptions );          // writes the data to a text file (see "text_io.cc" for function definition)
        else if ( this->outputFileType==112 )
            writeTextFile_gridIndex( dataToWrite, filename, variableName, userOptions );// writes the data to a text file, but on each line writes also the coordinates of the grid cell corresponding to the result being written (see "text_io.cc" for function definition)
        else if ( this->outputFileType==113 )
            writeTextFile_samplingPosition( dataToWrite, filename, variableName, userOptions ); // writes the data to a text file, but on each line writes also the sampling point coordinates corresponding to the result being written  (see "text_io.cc" for function definition)
        else if ( this->outputFileType==114 )
            writeTextFile_redshiftConePosition( dataToWrite, filename, variableName, userOptions ); // writes the data to a text file, but on each line writes also the sampling point coordinates for a redshift cone grid  (see "text_io.cc" for function definition)
        else if ( this->outputFileType==121 )
            writeMyFile( dataToWrite, filename, variableName, userOptions );   // the format that I use where it write to a binary file with a header that describes the data stored in it
        else if ( this->outputFileType==100 )
            writeSpecialFile( dataToWrite, filename, variableName, userOptions );   // the format that I use where it write to a binary file with a header that describes the data stored in it
    }
};


/*! This function writes the output data to a file, each different 'variable' being written to a separate file.
You can modify this function as you please. */
void writeOutputData(Quantities &uQuantities,
                     Quantities &aQuantities,
                     User_options const &userOptions)
{
    // select the file type for output
    OutputData output( userOptions.outputFileType );

    // output the desired quantities to file/files
    // outputs the density
    if ( userOptions.uField.density )
        output.write( uQuantities.density, userOptions.outputFilename + ".den", "density", userOptions );
    if ( userOptions.aField.density )
        output.write( aQuantities.density, userOptions.outputFilename + ".a_den", "volume averaged density", userOptions );

    // outputs the velocity
    if ( userOptions.uField.velocity )
        output.write( uQuantities.velocity, userOptions.outputFilename + ".vel", "velocity", userOptions );
    if ( userOptions.aField.velocity )
        output.write( aQuantities.velocity, userOptions.outputFilename + ".a_vel", "volume averaged velocity", userOptions );

    // outputs the velocity gradient
    if ( userOptions.uField.velocity_gradient )
        output.write( uQuantities.velocity_gradient, userOptions.outputFilename + ".velGrad", "velocity gradient", userOptions );
    if ( userOptions.aField.velocity_gradient )
        output.write( aQuantities.velocity_gradient, userOptions.outputFilename + ".a_velGrad", "volume averaged velocity gradient", userOptions );

    // outputs the velocity divergence
    if ( userOptions.uField.velocity_divergence )
        output.write( uQuantities.velocity_divergence, userOptions.outputFilename + ".velDiv", "velocity divergence", userOptions );
    if ( userOptions.aField.velocity_divergence )
        output.write( aQuantities.velocity_divergence, userOptions.outputFilename + ".a_velDiv", "volume averaged velocity divergence", userOptions );

    // outputs the velocity shear
    if ( userOptions.uField.velocity_shear )
        output.write( uQuantities.velocity_shear, userOptions.outputFilename + ".velShear", "velocity shear", userOptions );
    if ( userOptions.aField.velocity_shear )
        output.write( aQuantities.velocity_shear, userOptions.outputFilename + ".a_velShear", "volume averaged velocity shear", userOptions );

    // outputs the velocity vorticity
    if ( userOptions.uField.velocity_vorticity )
        output.write( uQuantities.velocity_vorticity, userOptions.outputFilename + ".velVort", "velocity vorticity", userOptions );
    if ( userOptions.aField.velocity_vorticity )
        output.write( aQuantities.velocity_vorticity, userOptions.outputFilename + ".a_velVort", "volume averaged velocity vorticity", userOptions );

    // outputs the velocity standard deviation
    if ( userOptions.aField.velocity_std )
        output.write( aQuantities.velocity_std, userOptions.outputFilename + ".a_velStd", "volume averaged velocity standard deviation", userOptions );

    // outputs the scalar fields
    if ( userOptions.uField.scalar )
        output.write( uQuantities.scalar, userOptions.outputFilename + ".scalar", "scalar", userOptions );
    if ( userOptions.aField.scalar )
        output.write( aQuantities.scalar, userOptions.outputFilename + ".a_scalar", "volume averaged scalar", userOptions );

    // outputs the scalar fields gradient
    if ( userOptions.uField.scalar_gradient )
        output.write( uQuantities.scalar_gradient, userOptions.outputFilename + ".scalarGrad", "scalar gradient", userOptions );
    if ( userOptions.aField.scalar_gradient )
        output.write( aQuantities.scalar_gradient, userOptions.outputFilename + ".a_scalarGrad", "volume averaged scalar gradient", userOptions );
}

#endif
